// chordsheetjs/lib/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CHORUS", {
  enumerable: true,
  get: function get() {
    return _constants.CHORUS;
  }
});
Object.defineProperty(exports, "Chord", {
  enumerable: true,
  get: function get() {
    return _chord["default"];
  }
});
Object.defineProperty(exports, "ChordLyricsPair", {
  enumerable: true,
  get: function get() {
    return _chord_lyrics_pair["default"];
  }
});
Object.defineProperty(exports, "ChordProFormatter", {
  enumerable: true,
  get: function get() {
    return _chord_pro_formatter["default"];
  }
});
Object.defineProperty(exports, "ChordProParser", {
  enumerable: true,
  get: function get() {
    return _chord_pro_parser["default"];
  }
});
Object.defineProperty(exports, "ChordSheetParser", {
  enumerable: true,
  get: function get() {
    return _chord_sheet_parser["default"];
  }
});
Object.defineProperty(exports, "ChordSheetSerializer", {
  enumerable: true,
  get: function get() {
    return _chord_sheet_serializer["default"];
  }
});
Object.defineProperty(exports, "Comment", {
  enumerable: true,
  get: function get() {
    return _comment["default"];
  }
});
Object.defineProperty(exports, "Composite", {
  enumerable: true,
  get: function get() {
    return _composite["default"];
  }
});
Object.defineProperty(exports, "HtmlDivFormatter", {
  enumerable: true,
  get: function get() {
    return _html_div_formatter["default"];
  }
});
Object.defineProperty(exports, "HtmlTableFormatter", {
  enumerable: true,
  get: function get() {
    return _html_table_formatter["default"];
  }
});
Object.defineProperty(exports, "INDETERMINATE", {
  enumerable: true,
  get: function get() {
    return _constants.INDETERMINATE;
  }
});
Object.defineProperty(exports, "Key", {
  enumerable: true,
  get: function get() {
    return _key["default"];
  }
});
Object.defineProperty(exports, "Line", {
  enumerable: true,
  get: function get() {
    return _line["default"];
  }
});
Object.defineProperty(exports, "Literal", {
  enumerable: true,
  get: function get() {
    return _literal["default"];
  }
});
Object.defineProperty(exports, "Metadata", {
  enumerable: true,
  get: function get() {
    return _metadata["default"];
  }
});
Object.defineProperty(exports, "NONE", {
  enumerable: true,
  get: function get() {
    return _constants.NONE;
  }
});
Object.defineProperty(exports, "NUMERIC", {
  enumerable: true,
  get: function get() {
    return _constants.NUMERIC;
  }
});
Object.defineProperty(exports, "Paragraph", {
  enumerable: true,
  get: function get() {
    return _paragraph["default"];
  }
});
Object.defineProperty(exports, "SYMBOL", {
  enumerable: true,
  get: function get() {
    return _constants.SYMBOL;
  }
});
Object.defineProperty(exports, "Song", {
  enumerable: true,
  get: function get() {
    return _song["default"];
  }
});
Object.defineProperty(exports, "TAB", {
  enumerable: true,
  get: function get() {
    return _constants.TAB;
  }
});
Object.defineProperty(exports, "Tag", {
  enumerable: true,
  get: function get() {
    return _tag["default"];
  }
});
Object.defineProperty(exports, "Ternary", {
  enumerable: true,
  get: function get() {
    return _ternary["default"];
  }
});
Object.defineProperty(exports, "TextFormatter", {
  enumerable: true,
  get: function get() {
    return _text_formatter["default"];
  }
});
Object.defineProperty(exports, "UltimateGuitarParser", {
  enumerable: true,
  get: function get() {
    return _ultimate_guitar_parser["default"];
  }
});
Object.defineProperty(exports, "VERSE", {
  enumerable: true,
  get: function get() {
    return _constants.VERSE;
  }
});
exports["default"] = void 0;
Object.defineProperty(exports, "parseChord", {
  enumerable: true,
  get: function get() {
    return _chord.parseChord;
  }
});

var _chord_pro_parser = _interopRequireDefault(require("./parser/chord_pro_parser"));

var _chord_sheet_parser = _interopRequireDefault(require("./parser/chord_sheet_parser"));

var _ultimate_guitar_parser = _interopRequireDefault(require("./parser/ultimate_guitar_parser"));

var _text_formatter = _interopRequireDefault(require("./formatter/text_formatter"));

var _html_table_formatter = _interopRequireDefault(require("./formatter/html_table_formatter"));

var _html_div_formatter = _interopRequireDefault(require("./formatter/html_div_formatter"));

var _chord_pro_formatter = _interopRequireDefault(require("./formatter/chord_pro_formatter"));

var _chord_lyrics_pair = _interopRequireDefault(require("./chord_sheet/chord_lyrics_pair"));

var _line = _interopRequireDefault(require("./chord_sheet/line"));

var _song = _interopRequireDefault(require("./chord_sheet/song"));

var _tag = _interopRequireDefault(require("./chord_sheet/tag"));

var _comment = _interopRequireDefault(require("./chord_sheet/comment"));

var _metadata = _interopRequireDefault(require("./chord_sheet/metadata"));

var _paragraph = _interopRequireDefault(require("./chord_sheet/paragraph"));

var _ternary = _interopRequireDefault(require("./chord_sheet/chord_pro/ternary"));

var _composite = _interopRequireDefault(require("./chord_sheet/chord_pro/composite"));

var _literal = _interopRequireDefault(require("./chord_sheet/chord_pro/literal"));

var _chord_sheet_serializer = _interopRequireDefault(require("./chord_sheet_serializer"));

var _constants = require("./constants");

var _chord = _interopRequireWildcard(require("./chord"));

var _key = _interopRequireDefault(require("./key"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _default = {
  ChordProParser: _chord_pro_parser["default"],
  ChordSheetParser: _chord_sheet_parser["default"],
  UltimateGuitarParser: _ultimate_guitar_parser["default"],
  TextFormatter: _text_formatter["default"],
  HtmlTableFormatter: _html_table_formatter["default"],
  HtmlDivFormatter: _html_div_formatter["default"],
  ChordProFormatter: _chord_pro_formatter["default"],
  ChordLyricsPair: _chord_lyrics_pair["default"],
  Line: _line["default"],
  Song: _song["default"],
  Tag: _tag["default"],
  Comment: _comment["default"],
  Metadata: _metadata["default"],
  Paragraph: _paragraph["default"],
  Ternary: _ternary["default"],
  Composite: _composite["default"],
  Literal: _literal["default"],
  ChordSheetSerializer: _chord_sheet_serializer["default"],
  CHORUS: _constants.CHORUS,
  INDETERMINATE: _constants.INDETERMINATE,
  TAB: _constants.TAB,
  VERSE: _constants.VERSE,
  NONE: _constants.NONE
};
exports["default"] = _default;
